sympy/matrices/expressions/matexpr.py

```diff
@@ -642,6 +655,17 @@ def _eval_derivative(self, v):
 
         return S.Zero
 
+    def _eval_subs(self, old, new):
+        # when replacing the symbol, the indices cannot
+        # access out of bounds
+        if old == self.symbol:
+            if old.func == new.func:
```

Comment:
Do you really want to compare `func` here, or are you actually looking for `type(old) == type(new)`?

Reply:
I don't have a nuanced understanding considering this:
```python
>>> Dummy().func
<class 'sympy.core.symbol.Dummy'>
>>> type(Dummy())
<class 'sympy.core.symbol.Dummy'>
```
So for me these seem like the same thing.

Reply:
Currently, `func` always returns the type, however, that means that `atom.func(*atom.args) == atom` generally can't work for `Atom`s (unless the atom stores no information) if `.func` always returns its type since it never stores it in `args`. So it should be made possible that `func` can also return something else.

I think the use of `func` as an alias for `type()` like this is more a remainder of python 2 style classes rather than something that should generally be prefered. If the question is whether it is the same class, then `type()` should be used since it is clearer to the general python reader and it would allow for a more flexible use of `func`.

Maybe @oscarbenjamin has some further insight.